File: sable-shared/src/crypto/hash.ts
// App-layer hashing — Node's `crypto` module wrapped for ergonomics.
//
// DB-layer hashing (token lookups, email_lookup) happens via the `digest()`
// SQL function inside the schema. App-layer hashing here is for cases where
// the Node service must compute the same hash *before* the query — e.g. the
// gateway computes `email_lookup = sha256(lower(email))` to look up a user at
// login, hashes a session token before INSERT, or verifies a webhook
// signature.
//
// Outputs are Buffers by default — matches the schema's BYTEA columns and
// avoids hex/base64 round-trips when writing to the DB.

import { createHash, createHmac, timingSafeEqual } from 'node:crypto';

// ---------------------------------------------------------------------------
// SHA-256
// ---------------------------------------------------------------------------

/** SHA-256 → Buffer (32 bytes). Matches BYTEA columns directly. */
export function sha256(input: string | Buffer | Uint8Array): Buffer {
  return createHash('sha256').update(input).digest();
}

/** SHA-256 → hex string. Use for logs / URL params; for DB writes use `sha256()`. */
export function sha256Hex(input: string | Buffer | Uint8Array): string {
  return createHash('sha256').update(input).digest('hex');
}

// ---------------------------------------------------------------------------
// HMAC-SHA-256 — service-to-service request signing, Stripe webhook verify
// ---------------------------------------------------------------------------

export function hmacSha256(key: string | Buffer, input: string | Buffer): Buffer {
  return createHmac('sha256', key).update(input).digest();
}

export function hmacSha256Hex(key: string | Buffer, input: string | Buffer): string {
  return createHmac('sha256', key).update(input).digest('hex');
}

// ---------------------------------------------------------------------------
// Comparison
// ---------------------------------------------------------------------------

/**
 * Constant-time equality check for hashes / HMACs / tokens. Required for any
 * comparison that an attacker might be able to time. Length mismatch returns
 * false (not throw) — the caller usually wants a boolean either way.
 */
export function constantTimeEqual(a: Buffer | Uint8Array, b: Buffer | Uint8Array): boolean {
  if (a.length !== b.length) return false;
  // timingSafeEqual requires same length (we already checked) and matching
  // ArrayBufferView types. Buffer.from with no copy when input is Buffer.
  const aBuf = Buffer.isBuffer(a) ? a : Buffer.from(a);
  const bBuf = Buffer.isBuffer(b) ? b : Buffer.from(b);
  return timingSafeEqual(aBuf, bBuf);
}

// ---------------------------------------------------------------------------
// Domain helpers
// ---------------------------------------------------------------------------

/**
 * Compute the email lookup hash. MUST match the schema's expression — both
 * sides do `trim` + `lower` + sha256 so a user pasting their email with
 * trailing whitespace doesn't silently fail login.
 *
 * Schema-side equivalent: `digest(lower(btrim($1)), 'sha256')`.
 * Change one, change both.
 */
export function emailLookup(email: string): Buffer {
  return sha256(email.trim().toLowerCase());
}

File: sable-shared-py/sable_shared/crypto/hash.py
"""App-layer hashing — stdlib ``hashlib`` / ``hmac`` wrapped for ergonomics.

DB-layer hashing (token lookups, email_lookup) happens via the
``digest()`` SQL function inside the schema. App-layer hashing here is
for cases where the Python service must compute the same hash *before*
the query — e.g. computing ``email_lookup = sha256(lower(email))`` for a
lookup, or verifying an HMAC on an inbound service-auth request.

Outputs are ``bytes`` by default — matches BYTEA columns directly and
avoids hex round-trips when writing to the DB.

Mirrors TS ``sable-shared/src/crypto/hash.ts``.
"""

from __future__ import annotations

import hashlib
import hmac


# ---------------------------------------------------------------------------
# SHA-256
# ---------------------------------------------------------------------------

def sha256(value: str | bytes) -> bytes:
    """SHA-256 → bytes (32). Matches BYTEA columns directly."""
    data = value.encode() if isinstance(value, str) else value
    return hashlib.sha256(data).digest()


def sha256_hex(value: str | bytes) -> str:
    """SHA-256 → hex string. Use for logs / URL params; for DB writes use ``sha256()``."""
    data = value.encode() if isinstance(value, str) else value
    return hashlib.sha256(data).hexdigest()


# ---------------------------------------------------------------------------
# HMAC-SHA-256 — service-to-service request signing, webhook verify
# ---------------------------------------------------------------------------

def hmac_sha256(key: str | bytes, message: str | bytes) -> bytes:
    key_bytes = key.encode() if isinstance(key, str) else key
    msg_bytes = message.encode() if isinstance(message, str) else message
    return hmac.new(key_bytes, msg_bytes, hashlib.sha256).digest()


def hmac_sha256_hex(key: str | bytes, message: str | bytes) -> str:
    return hmac_sha256(key, message).hex()


# ---------------------------------------------------------------------------
# Comparison
# ---------------------------------------------------------------------------

def constant_time_equal(a: bytes, b: bytes) -> bool:
    """Constant-time equality. Required for any hash/HMAC/token comparison
    an attacker could time. Length mismatch returns False (not throws)."""
    return hmac.compare_digest(a, b)


# ---------------------------------------------------------------------------
# Domain helpers
# ---------------------------------------------------------------------------

def email_lookup(email: str) -> bytes:
    """Compute the email lookup hash. MUST match the schema's expression —
    both sides do ``trim`` + ``lower`` + sha256 so a user pasting their
    email with trailing whitespace doesn't silently fail login.

    Schema-side equivalent: ``digest(lower(btrim($1)), 'sha256')``.
    Change one, change both.
    """
    return sha256(email.strip().lower())
